File: Arrays/1-two-sum/solution.dart
class Solution {
  List<int> twoSum(List<int> nums, int target) {
    final Map<int, int> map = {}; // value -> index

    for (int i = 0; i < nums.length; i++) {
        final int complement = target - nums[i];

        if (map.containsKey(complement)) {
        return [map[complement]!, i];
        }

        map[nums[i]] = i;
    }

    throw ArgumentError('No two sum solution found');
    }
}

File: Arrays/1-two-sum/solution.go
package main

import (
	"fmt"
)

func twoSum(nums []int, target int) []int {
	// Map to store number -> index
	numMap := make(map[int]int)

	for i, num := range nums {
		complement := target - num

		if idx, found := numMap[complement]; found {
			return []int{idx, i}
		}

		numMap[num] = i
	}

	// According to the problem, there is always exactly one solution.
	// Return nil or panic if no solution is found.
	return nil
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9

	result := twoSum(nums, target)
	fmt.Println(result) // Output: [0 1]
}
